src/tools/rest-backend/service_convert.cpp

```diff
@@ -0,0 +1,326 @@
+/**
+ * @file
+ *
+ * @brief implementation of the conversion service class
+ *
+ * @copyright BSD License (see doc/LICENSE.md or http://www.libelektra.org)
+ */
+
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/filesystem.hpp>
+
+#include <modules.hpp>
+#include <plugin.hpp>
+#include <plugindatabase.hpp>
+#include <pluginspec.hpp>
+#include <service.hpp>
+
+namespace kdbrest
+{
+
+namespace service
+{
+
+/**
+ * @brief tries to find a suitable plugin that can handle the supplied format.
+ *
+ * @param format a format that needs to be handled
+ *
+ * @return an object containing the format and the name of the suitable plugin
+ */
+model::PluginFormat ConvertEngine::findSuitablePlugin (const std::string & format)
+{
+	using namespace kdb;
+	using namespace kdb::tools;
+
+	ModulesPluginDatabase db;
+
+	try
+	{
+		PluginSpec plugin = db.lookupProvides (format);
+		std::string statusString = db.lookupInfo (plugin, "status");
+		std::vector<std::string> statuses;
+		boost::split (statuses, statusString, boost::is_any_of (" "));
+		return model::PluginFormat (format, plugin.getName (), statuses);
+	}
+	catch (kdb::tools::NoPlugin & e)
+	{
+		throw exception::UnsupportedConfigurationFormatException ();
+	}
+}
+
+/**
+ * @brief loads a list of enabled formats and plugins
+ * 
+ * Loads the configuration for the enabled formats from the
+ * key database, parses them and stores them as enabled formats.
+ *
+ * @return A vector containing the enabled convert formats with their
+ * corresponding plugins.
+ */
+std::vector<model::PluginFormat> ConvertEngine::loadEnabledFormats ()
+{
+	using namespace kdb;
+	using namespace kdb::tools;
+
+	ModulesPluginDatabase db;
+	std::vector<PluginSpec> plugins = db.lookupAllProvides ("storage");
+
+	// sort the plugins by status before processing
+	std::sort (plugins.begin (), plugins.end (), [&db](const PluginSpec & l, const PluginSpec & r) -> bool {
+		return db.calculateStatus (db.lookupInfo (l, "status")) > db.calculateStatus (db.lookupInfo (r, "status"));
+	});
+
+	std::vector<model::PluginFormat> result;
+	for (auto & plugin : plugins)
+	{
+		// find format
+		std::string provides = db.lookupInfo (plugin, "provides");
+		std::stringstream ss (provides);
+		std::string provider;
+		std::string format = "none";
+		while (ss >> provider)
+		{
+			if (boost::starts_with (provider, "storage/"))
+			{
+				format = provider.substr (8);
```

Comment:
avoid magic constants. Use something like `const char storage [] = "storage/"; const size_t sizeOfStorage = sizeof(storage)-1;` (if you meant that).